### PR TITLE
[Serializer] Updated the cookbook.

### PR DESCRIPTION
**Pull request by dunglas**
| Q | A |
| --- | --- |
| Doc fix? | yes |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | n/a |

### CODE REVIEW

File: cookbook/serializer.rst [hunk 1]
```diff
@@ -48,19 +46,44 @@ it in your configuration:
...
-                'enabled' => true
+                'enabled' => true,
             ),
         ));
 
+Using the Serializer Service
+---------------------------------
+
+Once enabled, the ``serializer`` service can be injected in any service where
+you need it used in a controller like the following::
```
Comment (xabbuh) [inline cookbook/serializer.rst]: [...] need it or it can be used [...]

File: cookbook/serializer.rst [hunk 2]
```diff
@@ -48,19 +46,44 @@ it in your configuration:
...
+
+    namespace AppBundle\Controller;
```
Comment (xabbuh) [inline cookbook/serializer.rst]:
```php
// src/AppBundle/Controller/DefaultController.php
```

must be added

File: cookbook/serializer.rst [hunk 3]
```diff
@@ -48,19 +46,44 @@ it in your configuration:
...
+
+            // ...
+        }
+    }
+
 Adding Normalizers and Encoders
 -------------------------------
 
+.. versionadded:: 2.7
+    :class:`Symfony\\Component\\Serializer\\Normalizer\\ObjectNormalizer` has
```
Comment (xabbuh) [inline cookbook/serializer.rst]: The [...]

File: cookbook/serializer.rst [hunk 4]
```diff
@@ -48,19 +46,44 @@ it in your configuration:
...
+    been enabled by default in Symfony 2.7. In prior versions you need to load
```
Comment (xabbuh) [inline cookbook/serializer.rst]: is enabled

File: cookbook/serializer.rst [hunk 5]
```diff
@@ -101,4 +124,96 @@ Here is an example on how to load the
...
         $container->setDefinition('get_set_method_normalizer', $definition);
 
-.. _JMSSerializerBundle: http://jmsyst.com/bundles/JMSSerializerBundle
+Using Serialization Groups Annotations
+--------------------------------------
+
+.. versionadded:: 2.7
+    The serialization group system has been added in Symfony 2.7.
+
+Enable :ref:`serialization groups annotation<component-serializer-attributes-groups>`
```
Comment (xabbuh) [inline cookbook/serializer.rst]: Please add a space between the label and the target.

File: cookbook/serializer.rst [hunk 6]
```diff
@@ -101,4 +124,96 @@ Here is an example on how to load the
...
+        ));
+
+Enabling the Metadata Cache
+---------------------------
+
+.. versionadded:: 2.7
+    Serializer metadata has been added in Symfony 2.7.
+
+Metadata used by the Serializer component such as groups can be cached to
+enhance application performance. Any service implementing ``Doctrine\Common\Cache\Cache``
```
Comment (xabbuh) [inline cookbook/serializer.rst]: implementing the

File: cookbook/serializer.rst [hunk 7]
```diff
@@ -101,4 +124,96 @@ Here is an example on how to load the
...
+interface can be used.
+
+A service leveraging `APCu`_ (and APC for PHP < 5.5) is builtin.
```
Comment (xabbuh) [inline cookbook/serializer.rst]: built-in

File: cookbook/serializer.rst [hunk 8]
```diff
@@ -101,4 +124,96 @@ Here is an example on how to load the
...
+            'serializer' => array(
+                'cache' => 'serializer.mapping.cache.apc',
+            ),
+        ));
+
+Going Further with the Serializer Component
+-------------------------------------------
+
+`DunglasApiBundle`_ provides an API system supporting `JSON-LD`_ and `Hydra Core Vocabulary`_
+hypermedia formats. It is built on top of the Symfony framework and its Serializer
```
Comment (xabbuh) [inline cookbook/serializer.rst]: Framework

File: cookbook/serializer.rst [hunk 9]
```diff
@@ -101,4 +124,96 @@ Here is an example on how to load the
...
+component.
+It provides custom normalizers and a custom encoder, custom metadata and a caching
```
Comment (xabbuh) [inline cookbook/serializer.rst]: There is no need for the newline here. You would have to add a blank line if you wanted to start a new paragraph.

**Comment by xabbuh** [comment]
Thanks to your work on this @dunglas! I left you some minor comments. Imho we should also add this to the configuration reference.

File: cookbook/serializer.rst [hunk 10]
```diff
@@ -48,19 +46,44 @@ it in your configuration:
         $container->loadFromExtension('framework', array(
             // ...
             'serializer' => array(
-                'enabled' => true
+                'enabled' => true,
             ),
         ));
 
+Using the Serializer Service
+---------------------------------
```
Comment (WouterJ) [inline cookbook/serializer.rst]: this tagline, as reSt calls it, should have exact the same length as the title text.

**Comment by dunglas** [comment]
@xabbuh and @WouterJ commend addressed.

File: cookbook/serializer.rst [hunk 11]
```diff
@@ -48,23 +46,50 @@ it in your configuration:
...
-                'enabled' => true
+                'enabled' => true,
             ),
         ));
 
+Using the Serializer Service
+----------------------------
+
+Once enabled, the ``serializer`` service can be injected in any service where
+you need or it can be used in a controller like the following::
```
Comment (xabbuh) [inline cookbook/serializer.rst]: [...] you need it or it [...]

File: cookbook/serializer.rst [hunk 12]
```diff
@@ -48,23 +46,50 @@ it in your configuration:
...
+
+    // src/AppBundle/Controller/DefaultController.php
+
```
Comment (xabbuh) [inline cookbook/serializer.rst]: This blank line must be removed.

File: cookbook/serializer.rst [hunk 13]
```diff
@@ -48,23 +46,50 @@ it in your configuration:
...
+    is enabled by default in Symfony 2.7. In prior versions you need to load
+    your own normalizer.
+
 Once enabled, the ``serializer`` service will be available in the container
 and will be loaded with two :ref:`encoders<component-serializer-encoders>`
 (:class:`Symfony\\Component\\Serializer\\Encoder\\JsonEncoder` and
 :class:`Symfony\\Component\\Serializer\\Encoder\\XmlEncoder`)
-but no :ref:`normalizers<component-serializer-normalizers>`, meaning you'll
-need to load your own.
+and the :class:`Symfony\\Component\\Serializer\\Normalizer\\ObjectNormalizer`
```
Comment (xabbuh) [inline cookbook/serializer.rst]: I would not use the `class` role here, but use the class name for the label of the reference too.

File: cookbook/serializer.rst [hunk 14]
```diff
@@ -101,4 +126,95 @@ Here is an example on how to load the
         $definition->addTag('serializer.normalizer');
         $container->setDefinition('get_set_method_normalizer', $definition);
 
-.. _JMSSerializerBundle: http://jmsyst.com/bundles/JMSSerializerBundle
+Using Serialization Groups Annotations
+--------------------------------------
+
+.. versionadded:: 2.7
+    The serialization group system has been added in Symfony 2.7.
```
Comment (xabbuh) [inline cookbook/serializer.rst]: I would reword this:

> Support for serialization groups was introduced in Symfony 2.7.

**Comment by dunglas** [comment]
@xabbuh done

File: components/serializer.rst [hunk 15]
```diff
@@ -476,7 +478,7 @@ There are several types of normalizers available:
 
 .. versionadded:: 2.6 The
     :class:`Symfony\\Component\\Serializer\\Normalizer\\PropertyNormalizer`
-    class was introduced in Symfony 2.6.
+    class was introduced in Symfony 2.  6.
```
Comment (xabbuh) [inline components/serializer.rst]: This looks weird. :)

File: cookbook/serializer.rst [hunk 16]
```diff
@@ -48,23 +46,48 @@ it in your configuration:
...
+            // ...
+        }
+    }
+
 Adding Normalizers and Encoders
 -------------------------------
 
+.. versionadded:: 2.7
+    The :class:`Symfony\\Component\\Serializer\\Normalizer\\ObjectNormalizer`
+    is enabled by default in Symfony 2.7. In prior versions you need to load
```
Comment (xabbuh) [inline cookbook/serializer.rst]: [...] version, you [...]

File: cookbook/serializer.rst [hunk 17]
```diff
@@ -48,23 +46,48 @@ it in your configuration:
...
+    your own normalizer.
+
 Once enabled, the ``serializer`` service will be available in the container
 and will be loaded with two :ref:`encoders<component-serializer-encoders>`
 (:class:`Symfony\\Component\\Serializer\\Encoder\\JsonEncoder` and
 :class:`Symfony\\Component\\Serializer\\Encoder\\XmlEncoder`)
-but no :ref:`normalizers<component-serializer-normalizers>`, meaning you'll
-need to load your own.
+and the :ref:`Symfony\\Component\\Serializer\\Normalizer\\ObjectNormalizer normalizer<component-serializer-normalizers>`.
```
Comment (xabbuh) [inline cookbook/serializer.rst]: Please move the `ref` role to its own line (you can append "and the" to the previous line).

File: cookbook/serializer.rst [hunk 18]
```diff
@@ -101,4 +124,95 @@ Here is an example on how to load the
...
         $container->setDefinition('get_set_method_normalizer', $definition);
 
-.. _JMSSerializerBundle: http://jmsyst.com/bundles/JMSSerializerBundle
+Using Serialization Groups Annotations
+--------------------------------------
+
+.. versionadded:: 2.7
+    Support for serialization groups was introduced in Symfony 2.7.
+
+Enable :ref:`serialization groups annotation<component-serializer-attributes-groups>`
```
Comment (xabbuh) [inline cookbook/serializer.rst]: missing space between label and reference

File: cookbook/serializer.rst (hunk 17 above)
Comment (xabbuh) [inline cookbook/serializer.rst]: and missing space between label and reference

**Comment by dunglas** [comment]
Should be ok now.

File: cookbook/serializer.rst [hunk 19]
```diff
@@ -48,23 +46,48 @@ it in your configuration:
...
+    your own normalizer.
+
 Once enabled, the ``serializer`` service will be available in the container
 and will be loaded with two :ref:`encoders<component-serializer-encoders>`
 (:class:`Symfony\\Component\\Serializer\\Encoder\\JsonEncoder` and
-:class:`Symfony\\Component\\Serializer\\Encoder\\XmlEncoder`)
-but no :ref:`normalizers<component-serializer-normalizers>`, meaning you'll
-need to load your own.
+:class:`Symfony\\Component\\Serializer\\Encoder\\XmlEncoder`) and the
+:ref:`Symfony\\Component\\Serializer\\Normalizer\\ObjectNormalizer normalizer <component-serializer-normalizers>`.
```
Comment (xabbuh) [inline cookbook/serializer.rst]: This will look a bit weird when being rendered. I would simply use "ObjectNormalizer" as the ref's label.

File: cookbook/serializer.rst [hunk 20]
```diff
@@ -101,4 +124,95 @@ Here is an example on how to load the
...
+            'serializer' => array(
+                'enable_annotations' => true,
+            ),
+        ));
+
+Enabling the Metadata Cache
+---------------------------
+
+.. versionadded:: 2.7
+    Serializer metadata has been added in Symfony 2.7.
```
Comment (xabbuh) [inline cookbook/serializer.rst]: [...] was introduced [...]

**Comment by xabbuh** [comment]
@dunglas Sorry, hopefully the last round of fixes.

**Comment by dunglas** [comment]
Should be fine now. #5332 is also finished.

**Comment by xabbuh** [comment]
:+1:

**Comment by weaverryan** [comment]
Awesome - thanks @dunglas! Minor tweaks at sha: f528471152d11beb52373b53a5fa0e914f230588
